File: Leetcode/PY/Leetcode260-只出现一次的数字2.py
from typing import List

# 位运算，找到不一样的位分别处理就行了
class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        mask = 0
        for num in nums:
            mask ^= num
        diff = 1
        while (diff & mask) == 0:
            diff <<= 1
        a,b = 0,0
        for num in nums:
            if (num & diff) != 0:
                a ^= num
            else:
                b ^= num
        return [a,b]

File: Leetcode/sbgo/Leetcode260-只出现一次的数字2.go
package sbgo

func singleNumber(nums []int) []int {
	mask, diff := 0, 1
	for _, num := range nums {
		mask ^= num
	}
	for (diff & mask) == 0 {
		diff <<= 1
	}
	a, b := 0, 0
	for _, num := range nums {
		if (diff & num) != 0 {
			a ^= num
		} else {
			b ^= num
		}
	}
	return []int{a, b}
}
